### PR TITLE
feat: always add ORDER BY with LIMIT or OFFSET

### PR DESCRIPTION
**Pull request by lohart13**
## Pull Request Checklist

- [x] Have you added new tests to prevent regressions?
- [ ] If a documentation update is necessary, have you opened a PR to [the documentation repository](https://github.com/sequelize/website/)? 
- [x] Did you update the typescript typings accordingly (if applicable)?
- [x] Does the description below contain a link to an existing issue (Closes #[issue]) or a description of the issue you are solving?
- [x] Does the name of your PR follow [our conventions](https://github.com/sequelize/sequelize/blob/main/CONTRIBUTING.md#6-commit-your-modifications)?

## Description Of Change

When using `OFFSET` and `LIMIT` (or `FETCH` depending on dialect), most dialects recommend including an `ORDER BY` clause to ensure that the results are returned in a consistent manner.

This PR ensures that whenever `options.limit` or `options.offset` is passed to the `selectQuery`, and `ORDER BY` clause is included in the query.

### CODE REVIEW

**Review by WikiRik — COMMENTED** [review]
Provisional review on the changes to limit/offset. Will review the changes related to the ORDER BY later

File: packages/core/src/dialects/snowflake/query-generator-typescript.ts [hunk 1]
```diff
@@ -135,4 +136,19 @@ export class SnowflakeQueryGeneratorTypeScript extends AbstractQueryGenerator {
   versionQuery() {
     return 'SELECT CURRENT_VERSION() AS "version"';
   }
+
+  protected _addLimitAndOffset(options: AddLimitOffsetOptions) {
+    let fragment = '';
+    if (options.limit != null) {
+      fragment += ` LIMIT ${this.escape(options.limit, options)}`;
+    } else if (options.offset) {
+      fragment += ` LIMIT NULL`;
```
[the remark this answers is not in the document]
Comment (lohart13) [reply packages/core/src/dialects/snowflake/query-generator-typescript.ts:145]: Snowflake still requires the `LIMIT` clause when `offset` is provided without a `limit` therefore I have opted to use `NULL` for readability.

File: packages/core/test/unit/query-generator/select-query.test.ts [hunk 2]
```diff
@@ -58,14 +59,75 @@ describe('QueryGenerator#selectQuery', () => {
...
+    expectsql(sql, {
+      mssql: `SELECT [id] FROM [Users] AS [User] ORDER BY [User].[id] OFFSET 1 ROWS FETCH NEXT 10 ROWS ONLY;`,
+      sqlite: 'SELECT `id` FROM `Users` AS `User` ORDER BY `User`.`id` LIMIT 10 OFFSET 1;',
+      'db2 ibmi': `SELECT "id" FROM "Users" AS "User" ORDER BY "User"."id" OFFSET 1 ROWS FETCH NEXT 10 ROWS ONLY;`,
+      'mariadb mysql': 'SELECT `id` FROM `Users` AS `User` ORDER BY `User`.`id` LIMIT 10 OFFSET 1;',
+      'postgres snowflake': 'SELECT "id" FROM "Users" AS "User" ORDER BY "User"."id" LIMIT 10 OFFSET 1;',
+    });
+  });
+
+  it('ignores 0 as offset', () => {
```
[the remark this answers is not in the document]
Comment (lohart13) [reply packages/core/test/unit/query-generator/select-query.test.ts:103]: For the dialects that support `LIMIT`/`FETCH` without an offset, when offset is 0, the `OFFSET` clause should not be included.
For instance, `OFFSET` is optional in `postgres`, `snowflake` however `mssql` requires `OFFSET` in order to use `FETCH`.
If you think this test isn't useful, I'll remove it.

[the remark this answers is not in the document]
Comment (lohart13) [reply packages/core/test/unit/query-generator/select-query.test.ts:103]: Good point, I'll add this in.

File: packages/core/src/dialects/abstract/query-generator.types.ts [hunk 3]
```diff
@@ -145,3 +146,9 @@ export interface QuoteTableOptions extends IndexHintable {
   alias: boolean | string;
   tableHints?: TableHints[];
 }
+
+export interface AddLimitOffsetOptions {
+  limit?: Nullish<number | BaseSqlExpression>;
+  offset?: Nullish<number | BaseSqlExpression>;
+  replacements?: BindOrReplacements;
```
[the remark this answers is not in the document]
Comment (lohart13) [reply packages/core/src/dialects/abstract/query-generator.types.ts:153]: There are some in `select-query.test.ts` where `limit` and `offset` are passed as literals and named replacements, let me know if this is enough or you want some more.

https://github.com/sequelize/sequelize/blob/11b960e95ab5b24743d50f850286e37612398ef0/packages/core/test/unit/query-generator/select-query.test.ts#L154-L565

[the remark this answers is not in the document]
Comment (WikiRik) [reply packages/core/src/dialects/abstract/query-generator.types.ts:153]: Ah yes, we indeed have those. But that's just literal. Do we support other expressions as well or should we narrow the type?

[the remark this answers is not in the document]
Comment (lohart13) [reply packages/core/src/dialects/abstract/query-generator.types.ts:153]: From what I understand, `LIMIT` and `OFFSET` accept integers, floats (only some dialects), begints and some strings like `NULL`, `ALL` (only some dialects).

If this is the case, limiting the type to number and Literal should provide enough options for user input.

File: packages/core/test/unit/query-generator/select-query.test.ts [hunk 4]
```diff
@@ -58,14 +59,75 @@ describe('QueryGenerator#selectQuery', () => {
     }, User);
```
[the remark this answers is not in the document]
Comment (lohart13) [reply packages/core/test/unit/query-generator/select-query.test.ts]: Added now

File: packages/core/src/dialects/postgres/query-generator-typescript.ts [hunk 5]
```diff
@@ -209,4 +210,19 @@ export class PostgresQueryGeneratorTypeScript extends AbstractQueryGenerator {
   versionQuery() {
     return 'SHOW SERVER_VERSION';
   }
+
+  protected _addLimitAndOffset(options: AddLimitOffsetOptions) {
+    let fragment = '';
+    if (options.limit != null) {
+      fragment += ` LIMIT ${this.escape(options.limit, options)}`;
+    } else if (options.offset) {
+      fragment += ` LIMIT NULL`;
```
[the remark this answers is not in the document]
Comment (lohart13) [reply packages/core/src/dialects/postgres/query-generator-typescript.ts]: Updated now to only include `LIMIT` if `options.limit` has a value.

File: packages/core/src/dialects/abstract/query-generator.types.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (lohart13) [reply packages/core/src/dialects/abstract/query-generator.types.ts]: Types updated now

File: packages/core/test/unit/query-generator/select-query.test.ts [hunk 6]
```diff
@@ -58,14 +59,75 @@ describe('QueryGenerator#selectQuery', () => {
...
+  it('support limit without offset', () => {
+    const sql = queryGenerator.selectQuery(User.table, {
+      model: User,
+      attributes: ['id'],
+      limit: 10,
+    }, User);
+
+    expectsql(sql, {
+      mssql: `SELECT [id] FROM [Users] AS [User] ORDER BY [User].[id] OFFSET 0 ROWS FETCH NEXT 10 ROWS ONLY;`,
+      sqlite: 'SELECT `id` FROM `Users` AS `User` ORDER BY `User`.`id` LIMIT 10;',
```
[the remark this answers is not in the document]
Comment (lohart13) [reply packages/core/test/unit/query-generator/select-query.test.ts]: Done.

File: packages/core/test/unit/sql/offset-limit.test.js [hunk 7]
```diff
@@ -1,104 +0,0 @@
...
-        ['email', 'DESC'], // for MSSQL
-      ],
-    }, {
-      default: ` LIMIT 10 OFFSET ''';DELETE FROM user'`,
-      mariadb: ` LIMIT 10 OFFSET '\\';DELETE FROM user'`,
-      mysql: ` LIMIT 10 OFFSET '\\';DELETE FROM user'`,
-      db2: ` OFFSET ''';DELETE FROM user' ROWS FETCH NEXT 10 ROWS ONLY`,
-      mssql: ` OFFSET N''';DELETE FROM user' ROWS FETCH NEXT 10 ROWS ONLY`,
-      ibmi: ` OFFSET ''';DELETE FROM user' ROWS FETCH NEXT 10 ROWS ONLY`,
-    });
```
[the remark this answers is not in the document]
Comment (lohart13) [reply packages/core/test/unit/sql/offset-limit.test.js:91]: I've moved the tests there now.

File: packages/core/test/unit/query-generator/select-query.test.ts [hunk 8]
```diff
@@ -50,22 +51,123 @@ describe('QueryGenerator#selectQuery', () => {
...
-      snowflake: 'SELECT "id" FROM "Users" AS "User" LIMIT NULL OFFSET 1;',
-      db2: `SELECT "id" FROM "Users" AS "User" OFFSET 1 ROWS;`,
-      ibmi: 'SELECT "id" FROM "Users" AS "User" OFFSET 1 ROWS',
-      mssql: `SELECT [id] FROM [Users] AS [User] ORDER BY [User].[id] OFFSET 1 ROWS;`,
+      expectsql(sql, {
+        mssql: `SELECT [id] FROM [Users] AS [User] ORDER BY [User].[id] OFFSET 1 ROWS;`,
+        sqlite: 'SELECT `id` FROM `Users` AS `User` ORDER BY `User`.`id` LIMIT -1 OFFSET 1;',
+        postgres: 'SELECT "id" FROM "Users" AS "User" ORDER BY "User"."id" OFFSET 1;',
+        snowflake: 'SELECT "id" FROM "Users" AS "User" ORDER BY "User"."id" LIMIT NULL OFFSET 1;',
+        'db2 ibmi': `SELECT "id" FROM "Users" AS "User" ORDER BY "User"."id" OFFSET 1 ROWS;`,
```
Comment (WikiRik) [inline packages/core/test/unit/query-generator/select-query.test.ts:67]:
```suggestion
        'mssql db2 ibmi': `SELECT [id] FROM [Users] AS [User] ORDER BY [User].[id] OFFSET 1 ROWS;`,
        sqlite: 'SELECT `id` FROM `Users` AS `User` ORDER BY `User`.`id` LIMIT -1 OFFSET 1;',
        postgres: 'SELECT "id" FROM "Users" AS "User" ORDER BY "User"."id" OFFSET 1;',
        snowflake: 'SELECT "id" FROM "Users" AS "User" ORDER BY "User"."id" LIMIT NULL OFFSET 1;',
```

File: packages/core/test/unit/query-generator/select-query.test.ts [hunk 9]
```diff
@@ -50,22 +51,123 @@ describe('QueryGenerator#selectQuery', () => {
...
+        model: User,
+        attributes: ['id'],
+        offset: 1,
+        limit: 10,
+      }, User);
+
+      expectsql(sql, {
+        default: 'SELECT [id] FROM [Users] AS [User] ORDER BY [User].[id] LIMIT 10 OFFSET 1;',
+        mssql: `SELECT [id] FROM [Users] AS [User] ORDER BY [User].[id] OFFSET 1 ROWS FETCH NEXT 10 ROWS ONLY;`,
+        'db2 ibmi': `SELECT "id" FROM "Users" AS "User" ORDER BY "User"."id" OFFSET 1 ROWS FETCH NEXT 10 ROWS ONLY;`,
```
Comment (WikiRik) [inline packages/core/test/unit/query-generator/select-query.test.ts:97]:
```suggestion
        'mssql db2 ibmi': `SELECT [id] FROM [Users] AS [User] ORDER BY [User].[id] OFFSET 1 ROWS FETCH NEXT 10 ROWS ONLY;`,
```

File: packages/core/test/unit/query-generator/select-query.test.ts [hunk 10]
```diff
@@ -50,22 +51,123 @@ describe('QueryGenerator#selectQuery', () => {
...
+        attributes: ['id'],
+        // @ts-expect-error -- testing invalid limit
+        limit: `';DELETE FROM user`,
+      }, User);
+
+      expectsql(sql, {
+        default: `SELECT [id] FROM [Users] AS [User] ORDER BY [User].[id] LIMIT ''';DELETE FROM user';`,
+        mssql: `SELECT [id] FROM [Users] AS [User] ORDER BY [User].[id] OFFSET 0 ROWS FETCH NEXT N''';DELETE FROM user' ROWS ONLY;`,
+        'db2 ibmi': `SELECT "id" FROM "Users" AS "User" ORDER BY "User"."id" FETCH NEXT ''';DELETE FROM user' ROWS ONLY;`,
+        'mariadb mysql': 'SELECT `id` FROM `Users` AS `User` ORDER BY `User`.`id` LIMIT \'\\\';DELETE FROM user\';',
```
Comment (WikiRik) [inline packages/core/test/unit/query-generator/select-query.test.ts:152]:
```suggestion
```
This is the same as default, right?

File: packages/core/test/unit/query-generator/select-query.test.ts [hunk 11]
```diff
@@ -50,22 +51,123 @@ describe('QueryGenerator#selectQuery', () => {
...
+        limit: 10,
+        // @ts-expect-error -- testing invalid offset
+        offset: `';DELETE FROM user`,
+      }, User);
+
+      expectsql(sql, {
+        default: `SELECT [id] FROM [Users] AS [User] ORDER BY [User].[id] LIMIT 10 OFFSET ''';DELETE FROM user';`,
+        mssql: `SELECT [id] FROM [Users] AS [User] ORDER BY [User].[id] OFFSET N''';DELETE FROM user' ROWS FETCH NEXT 10 ROWS ONLY;`,
+        'db2 ibmi': `SELECT "id" FROM "Users" AS "User" ORDER BY "User"."id" OFFSET ''';DELETE FROM user' ROWS FETCH NEXT 10 ROWS ONLY;`,
+        'mariadb mysql': 'SELECT `id` FROM `Users` AS `User` ORDER BY `User`.`id` LIMIT 10 OFFSET \'\\\';DELETE FROM user\';',
```
Comment (WikiRik) [inline packages/core/test/unit/query-generator/select-query.test.ts:169]:
```suggestion
```

**Review by WikiRik — APPROVED** [review]
Four suggestions for the unit tests. I haven't run them before making it, so please check before you commit. They're not blocking so I'll approve either way.

Thanks for working on this!

File: packages/core/test/unit/query-generator/select-query.test.ts (hunk 10 above)
Comment (lohart13) [reply packages/core/test/unit/query-generator/select-query.test.ts:152]: No, it's not. MariaDB and MySQL uses a backslash to escape the single quote.

File: packages/core/test/unit/query-generator/select-query.test.ts (hunk 11 above)
Comment (lohart13) [reply packages/core/test/unit/query-generator/select-query.test.ts:169]: MariaDB and MySQL uses a backslash to escape the single quote.

**Comment by lohart13** [comment]
Not sure a documentation update is required for this change but will tag to https://github.com/sequelize/website/issues/501

**Comment by ephys** [comment]
I haven't checked if this was fixed yet but I think this PR introduced a bug for queries that use `GROUP BY`

**Pull request merged**